### PR TITLE
tech-debt/aws_instance: add state conf for root device and metadata updates

### PR DESCRIPTION
**Pull request by anGie44**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates https://github.com/terraform-providers/terraform-provider-aws/issues/14383

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
tech-debt/aws_instance: add state conf handling for root device.0.delete_on_termination and metadata updates
```

Output from acceptance testing in TeamCity (these generally pass locally):

```
--- PASS: TestAccAWSInstance_EbsRootDevice_ModifyDeleteOnTermination (117.16s)
--- PASS: TestAccAWSInstance_EbsRootDevice_MultipleBlockDevices_ModifyDeleteOnTermination (130.30s)
--- PASS: TestAccAWSInstance_EbsRootDevice_ModifyAll (127.43s)
--- PASS: TestAccAWSInstance_metadataOptions (150.91s)
```

### CODE REVIEW

File: aws/resource_aws_instance.go [hunk 1]
```diff
@@ -1341,6 +1341,21 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
 				if err != nil {
 					return fmt.Errorf("Error updating metadata options: %s", err)
 				}
+				stateConf := &resource.StateChangeConf{
+					Pending:    []string{ec2.InstanceMetadataOptionsStatePending},
+					Target:     []string{ec2.InstanceMetadataOptionsStateApplied, ec2.InstanceStateNameRunning},
```
Comment (anGie44) [inline aws/resource_aws_instance.go:1346]: API docs indicate "applied" state follows the "pending" state returned from `ModifyInstanceMetadataOptions`  but in testing this, "applied" wasn't recognized as a valid target state, only "running" ..perhaps this should just be kept as "running"?

Comment (anGie44) [reply aws/resource_aws_instance.go:1346]: would we consider moving these statechangeconf references to a service package like the ec2 waiter/status ones as part of this tech-debt item? looking at the resource, we have quite a number of references that maybe could be consolidated

Comment (bflad) [reply aws/resource_aws_instance.go:1346]: We might need a separate refresh function for this as those enumerations appear to be related to the `InstanceMetadataOptionsResponse` type:

```go
// The metadata options for the instance.
type InstanceMetadataOptionsResponse struct {
// ...
	// The state of the metadata option changes.
	//
	// pending - The metadata options are being updated and the instance is not
	// ready to process metadata traffic with the new selection.
	//
	// applied - The metadata options have been successfully applied on the instance.
	State *string `locationName:"state" type:"string" enum:"InstanceMetadataOptionsState"`
}
```

Which can be found in the `Instance.MetadataOptions` field. 👍 

If we are adding new functions, I think adding them to the internal package is fine and we can worry about refactoring existing ones separately.

Comment (bflad) [reply aws/resource_aws_instance.go:1346]: As for the block device changes, there's also a `Status` field buried deep in `Instance.BlockDeviceMappings[].Ebs.Status`, but not sure how well that aligns with things like updating delete on termination and such, as its enumeration is:

```go
func AttachmentStatus_Values() []string {
	return []string{
		AttachmentStatusAttaching,
		AttachmentStatusAttached,
		AttachmentStatusDetaching,
		AttachmentStatusDetached,
	}
}
```

Comment (anGie44) [reply aws/resource_aws_instance.go]: @bflad the enumerations here didn't seem to work as we'd hope so instead I'm experimenting with an updated refresh func in 9f46770 to check for the actual `deleteOnTermination` value. would that be more in line with what we this type of update?

**Review by bflad — APPROVED** [review]
Looks great 🚀 

Output from acceptance testing:

```
--- FAIL: TestAccAWSInstance_instanceProfileChange (123.29s) # failure unrelated
--- PASS: TestAccAWSInstance_addSecondaryInterface (150.81s)
--- PASS: TestAccAWSInstance_addSecurityGroupNetworkInterface (167.76s)
--- PASS: TestAccAWSInstance_associatePublic_defaultPrivate (97.92s)
--- PASS: TestAccAWSInstance_associatePublic_defaultPublic (114.38s)
--- PASS: TestAccAWSInstance_associatePublic_explicitPrivate (102.47s)
--- PASS: TestAccAWSInstance_associatePublic_explicitPublic (114.23s)
--- PASS: TestAccAWSInstance_associatePublic_overridePrivate (82.45s)
--- PASS: TestAccAWSInstance_associatePublic_overridePublic (95.17s)
--- PASS: TestAccAWSInstance_associatePublicIPAndPrivateIP (110.80s)
--- PASS: TestAccAWSInstance_atLeastOneOtherEbsVolume (187.36s)
--- PASS: TestAccAWSInstance_basic (78.43s)
--- PASS: TestAccAWSInstance_blockDevices (94.04s)
--- PASS: TestAccAWSInstance_changeInstanceType (174.07s)
--- PASS: TestAccAWSInstance_CreditSpecification_Empty_NonBurstable (121.90s)
--- PASS: TestAccAWSInstance_creditSpecification_isNotAppliedToNonBurstable (108.89s)
--- PASS: TestAccAWSInstance_creditSpecification_standardCpuCredits (114.51s)
--- PASS: TestAccAWSInstance_creditSpecification_standardCpuCredits_t2Tot3Taint (391.27s)
--- PASS: TestAccAWSInstance_creditSpecification_unknownCpuCredits_t2 (102.28s)
--- PASS: TestAccAWSInstance_creditSpecification_unknownCpuCredits_t3 (303.78s)
--- PASS: TestAccAWSInstance_creditSpecification_unlimitedCpuCredits (133.15s)
--- PASS: TestAccAWSInstance_creditSpecification_unlimitedCpuCredits_t2Tot3Taint (368.61s)
--- PASS: TestAccAWSInstance_creditSpecification_unspecifiedDefaultsToStandard (83.89s)
--- PASS: TestAccAWSInstance_CreditSpecification_UnspecifiedToEmpty_NonBurstable (107.90s)
--- PASS: TestAccAWSInstance_creditSpecification_updateCpuCredits (145.32s)
--- PASS: TestAccAWSInstance_creditSpecificationT3_standardCpuCredits (91.15s)
--- PASS: TestAccAWSInstance_creditSpecificationT3_unlimitedCpuCredits (119.55s)
--- PASS: TestAccAWSInstance_creditSpecificationT3_unspecifiedDefaultsToUnlimited (122.18s)
--- PASS: TestAccAWSInstance_creditSpecificationT3_updateCpuCredits (154.07s)
--- PASS: TestAccAWSInstance_disableApiTermination (130.53s)
--- PASS: TestAccAWSInstance_disappears (80.99s)
--- PASS: TestAccAWSInstance_EbsBlockDevice_InvalidIopsForVolumeType (13.54s)
--- PASS: TestAccAWSInstance_EbsBlockDevice_KmsKeyArn (113.99s)
--- PASS: TestAccAWSInstance_EbsRootDevice_basic (122.38s)
--- PASS: TestAccAWSInstance_EbsRootDevice_ModifyAll (154.17s)
--- PASS: TestAccAWSInstance_EbsRootDevice_ModifyDeleteOnTermination (128.92s)
--- PASS: TestAccAWSInstance_EbsRootDevice_ModifyIOPS_Io1 (142.33s)
--- PASS: TestAccAWSInstance_EbsRootDevice_ModifyIOPS_Io2 (119.25s)
--- PASS: TestAccAWSInstance_EbsRootDevice_ModifySize (167.24s)
--- PASS: TestAccAWSInstance_EbsRootDevice_ModifyType (125.74s)
--- PASS: TestAccAWSInstance_EbsRootDevice_MultipleBlockDevices_ModifyDeleteOnTermination (146.53s)
--- PASS: TestAccAWSInstance_EbsRootDevice_MultipleBlockDevices_ModifySize (145.85s)
--- PASS: TestAccAWSInstance_EbsRootDevice_MultipleDynamicEBSBlockDevices (196.22s)
--- PASS: TestAccAWSInstance_Empty_PrivateIP (110.62s)
--- PASS: TestAccAWSInstance_forceNewAndTagsDrift (154.39s)
--- PASS: TestAccAWSInstance_getPasswordData_falseToTrue (131.51s)
--- PASS: TestAccAWSInstance_getPasswordData_trueToFalse (204.70s)
--- PASS: TestAccAWSInstance_GP2IopsDevice (88.44s)
--- PASS: TestAccAWSInstance_GP2WithIopsValue (10.08s)
--- PASS: TestAccAWSInstance_hibernation (162.29s)
--- PASS: TestAccAWSInstance_inDefaultVpcBySgId (84.76s)
--- PASS: TestAccAWSInstance_inDefaultVpcBySgName (96.09s)
--- PASS: TestAccAWSInstance_ipv6_supportAddressCount (92.14s)
--- PASS: TestAccAWSInstance_ipv6_supportAddressCountWithIpv4 (100.52s)
--- PASS: TestAccAWSInstance_ipv6AddressCountAndSingleAddressCausesError (12.13s)
--- PASS: TestAccAWSInstance_keyPairCheck (77.39s)
--- PASS: TestAccAWSInstance_metadataOptions (123.01s)
--- PASS: TestAccAWSInstance_multipleRegions (185.73s)
--- PASS: TestAccAWSInstance_NetworkInstanceRemovingAllSecurityGroups (104.74s)
--- PASS: TestAccAWSInstance_NetworkInstanceSecurityGroups (95.40s)
--- PASS: TestAccAWSInstance_NetworkInstanceVPCSecurityGroupIDs (97.83s)
--- PASS: TestAccAWSInstance_NewNetworkInterface_EmptyPrivateIPAndSecondaryPrivateIPs (96.88s)
--- PASS: TestAccAWSInstance_NewNetworkInterface_EmptyPrivateIPAndSecondaryPrivateIPsUpdate (160.07s)
--- PASS: TestAccAWSInstance_NewNetworkInterface_PrivateIPAndSecondaryPrivateIPs (95.81s)
--- PASS: TestAccAWSInstance_NewNetworkInterface_PrivateIPAndSecondaryPrivateIPsUpdate (114.40s)
--- PASS: TestAccAWSInstance_NewNetworkInterface_PublicIPAndSecondaryPrivateIPs (416.46s)
--- PASS: TestAccAWSInstance_noAMIEphemeralDevices (109.32s)
--- PASS: TestAccAWSInstance_placementGroup (76.53s)
--- PASS: TestAccAWSInstance_primaryNetworkInterface (105.25s)
--- PASS: TestAccAWSInstance_primaryNetworkInterfaceSourceDestCheck (103.50s)
--- PASS: TestAccAWSInstance_privateIP (100.18s)
--- PASS: TestAccAWSInstance_RootBlockDevice_KmsKeyArn (68.88s)
--- PASS: TestAccAWSInstance_rootBlockDeviceMismatch (84.07s)
--- PASS: TestAccAWSInstance_rootInstanceStore (82.49s)
--- PASS: TestAccAWSInstance_sourceDestCheck (168.88s)
--- PASS: TestAccAWSInstance_tags (117.66s)
--- PASS: TestAccAWSInstance_UserData_EmptyStringToUnspecified (107.56s)
--- PASS: TestAccAWSInstance_UserData_UnspecifiedToEmptyString (99.16s)
--- PASS: TestAccAWSInstance_userDataBase64 (131.12s)
--- PASS: TestAccAWSInstance_volumeTags (146.13s)
--- PASS: TestAccAWSInstance_volumeTagsComputed (116.41s)
--- PASS: TestAccAWSInstance_vpc (101.62s)
--- PASS: TestAccAWSInstance_withIamInstanceProfile (129.34s)
--- SKIP: TestAccAWSInstance_inEc2Classic (1.15s)
--- SKIP: TestAccAWSInstance_outpost (1.80s)

--- PASS: TestAccAWSInstanceDataSource_AzUserData (109.33s)
--- PASS: TestAccAWSInstanceDataSource_basic (113.79s)
--- PASS: TestAccAWSInstanceDataSource_blockDevices (91.72s)
--- PASS: TestAccAWSInstanceDataSource_creditSpecification (119.07s)
--- PASS: TestAccAWSInstanceDataSource_EbsBlockDevice_KmsKeyId (131.39s)
--- PASS: TestAccAWSInstanceDataSource_getPasswordData_falseToTrue (145.73s)
--- PASS: TestAccAWSInstanceDataSource_getPasswordData_trueToFalse (169.50s)
--- PASS: TestAccAWSInstanceDataSource_GetUserData (139.89s)
--- PASS: TestAccAWSInstanceDataSource_GetUserData_NoUserData (132.48s)
--- PASS: TestAccAWSInstanceDataSource_gp2IopsDevice (106.67s)
--- PASS: TestAccAWSInstanceDataSource_keyPair (117.32s)
--- PASS: TestAccAWSInstanceDataSource_metadataOptions (86.58s)
--- PASS: TestAccAWSInstanceDataSource_PlacementGroup (93.43s)
--- PASS: TestAccAWSInstanceDataSource_privateIP (100.05s)
--- PASS: TestAccAWSInstanceDataSource_RootBlockDevice_KmsKeyId (88.80s)
--- PASS: TestAccAWSInstanceDataSource_rootInstanceStore (294.95s)
--- PASS: TestAccAWSInstanceDataSource_secondaryPrivateIPs (111.06s)
--- PASS: TestAccAWSInstanceDataSource_SecurityGroups (109.58s)
--- PASS: TestAccAWSInstanceDataSource_tags (110.07s)
--- PASS: TestAccAWSInstanceDataSource_VPC (111.10s)
--- PASS: TestAccAWSInstanceDataSource_VPCSecurityGroups (117.06s)
--- PASS: TestAccAWSInstancesDataSource_basic (76.95s)
--- PASS: TestAccAWSInstancesDataSource_instance_state_names (77.72s)
--- PASS: TestAccAWSInstancesDataSource_tags (234.57s)
```

**Pull request merged**